### PR TITLE
[Bugfix] Mask OOV tokens IDs from the Rejection Sampler

### PR DESCRIPTION
**Pull request by gnovack**
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose
The `sample_recovered_tokens_kernel` in the V1 Rejection Sampler currently returns out-of-vocabulary tokens when the input `target_probs` contains `nan` values. This sampling of OOV token ids causes the engine to crash as described in #20531

This seems to be because the kernel pads the vocabulary size with `-inf` and Triton's argmax function selects the index of the first `-inf` when the previous elements in the tensor are `nan` (i.e. `tl.argmax([nan, nan, -inf, ...]) --> 2`).

This PR masks out any OOV token ids so that any output token which is outside the vocabulary will be set to 0. 

The approach taken here is admittedly a band-aid, as it merely aligns the Rejection Sampler's behavior with that of the base target model sampler (i.e. by sampling token id `0` when the logits are full of `nan` values). This spitting out of `0` tokens when the logits have `nan`s, however, seems to be more of an accident, as opposed to an intended behavior; so I am curious to hear if there are any other thoughts on how to better handle this scenario.

## Test Plan

Added a new test case, `test_nan_logits`, which illustrates the behavior described above. 

## Test Result

This test fails without the change applied in this PR, and now passes.

## (Optional) Documentation Update

N/A

### CODE REVIEW

**Comment by ekagra-ranjan** [comment]
@gnovack - thanks for your PR. I was wondering if this fixed your issue?

File: tests/v1/sample/test_rejection_sampler.py [hunk 1]
```diff
@@ -198,6 +198,33 @@ def test_empty_sequence(rejection_sampler):
     assert torch.equal(output, expected)
 
 
+def test_nan_logits(rejection_sampler):
+    """Test handling nan values in the target logits"""
+    spec_tokens: list[list[int]] = [[0]]
+    output_tokens = [[0, 0]]  # Just the bonus token
```
Comment (ekagra-ranjan) [inline tests/v1/sample/test_rejection_sampler.py:204]: right now, all the tokens are 0 which gives the impression that they are related in different places. However, they are not. Can we instead change it to
```
spec_tokens: list[list[int]] = [[1]]
output_tokens = [[2, 3]]  # Just the bonus token
```

The bonus token defined later doesn't matter since the draft token will be rejected and be sampled as 0 so the above non-0 token ids will make this apparent. 

IIUC, no matter what the spec_tokens and output_tokens and bonus tokens are, if the logits are nan then it will always return `[0, PLACEHOLDER_TOKEN_ID]`

File: vllm/v1/sample/rejection_sampler.py [hunk 2]
```diff
@@ -425,6 +425,9 @@ def sample_recovered_tokens(
         triton.next_power_of_2(vocab_size),
         NO_DRAFT_PROBS=draft_probs is None,
     )
+    # mask any out-of-vocabulary values in recovered_token_ids
+    recovered_token_ids = recovered_token_ids * (recovered_token_ids
+                                                 < vocab_size)
```
[the remark this answers is not in the document]
Comment (ekagra-ranjan) [reply vllm/v1/sample/rejection_sampler.py:430]: Adding this to the kernel makes sense. 
Could you also add a comment as to why we replace the OOV with 0 for posterity?

**Review by ekagra-ranjan — APPROVED** [review]
LGTM. Left some comments. We should get this in.

File: vllm/v1/sample/rejection_sampler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xyang16) [reply vllm/v1/sample/rejection_sampler.py:430]: The mask value of q can also be set to nan, so that it can guarantee the OOV index will never be picked as recovered_id.

```
    q = tl.load(q_ptr + req_idx * vocab_size + vocab_offset,
                mask=vocab_offset < vocab_size,
                other=float("nan"))
```

[the remark this answers is not in the document]
Comment (xyang16) [reply vllm/v1/sample/rejection_sampler.py:430]: The OOV value of `q` can also be set to `nan` instead of `-inf`, so that it can guarantee the OOV index will never be picked as `recovered_id`.

```
    q = tl.load(q_ptr + req_idx * vocab_size + vocab_offset,
                mask=vocab_offset < vocab_size,
                other=float("nan"))
```
